osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs

```diff
@@ -52,23 +54,36 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool hidd
                 var currentObj = (OsuDifficultyHitObject)current.Previous(i);
                 var currentHitObject = (OsuHitObject)(currentObj.BaseObject);
 
+                cumulativeStrainTime += lastObj.StrainTime;
+
                 if (!(currentObj.BaseObject is Spinner))
                 {
-                    double jumpDistance = (osuHitObject.StackedPosition - currentHitObject.StackedEndPosition).Length;
+                    double pixelJumpDistance = (osuHitObject.StackedPosition - currentHitObject.StackedEndPosition).Length;
+                    double flashlightRadius = getComboScaleFor(currentObj.CurrentMaxCombo);
+                    double objectOpacity = osuCurrent.OpacityAt(currentHitObject.StartTime, hidden);
 
-                    cumulativeStrainTime += lastObj.StrainTime;
+                    // Consider the jump from the lazy end position for sliders.
+                    if (currentHitObject is Slider currentSlider)
+                    {
+                        Vector2 lazyEndPosition = currentSlider.LazyEndPosition ?? currentSlider.StackedPosition;
+                        pixelJumpDistance = Math.Min(pixelJumpDistance, (osuHitObject.StackedPosition - lazyEndPosition).Length);
+                    }
 
-                    // We want to nerf objects that can be easily seen within the Flashlight circle radius.
+                    // Apply a nerf based on the visibility from the current object.
+                    double distanceNerf = Math.Min(1.0, pixelJumpDistance / (flashlightRadius + osuHitObject.Radius - 80));
```

Comment:
Does the `osuHitObject.Radius` term here make much sense? The flashlight radius doesn't change depending on circle size.

Also, I would quickly take a look at L77-78, because the `smallDistNerf` is really *really* similar to the `distanceNerf`: https://www.desmos.com/calculator/nm3cddxy4m

The difference being that... it only runs when `i == 0`, which means it only runs once in the loop? Which doesn't seem accurate given it's purpose.

Reply:
`osuHitObject.Radius` in the `distanceNerf` is used to calculate the note's visibility, since bigger circles are gonna poke into the flashlight radius more. This ends up nerfing smaller circles inside the radius, and nerfs bigger notes just outside the radius. This is only used to calculate the note's total visibility in `visibilityNerf`, so perhaps a different name like `distanceVisibility` is better?

As for `smallDistNerf` being such a similiar formula, I tried using `distanceNerf` for it but it gives worse results. The reason it's only calculated for `i == 0` is to consider stream maps so they don't blow up.